File: app/src/main/java/com/softwareoverflow/HangTight/workout/WorkoutTimer.java
package com.softwareoverflow.HangTight.workout;

import android.os.CountDownTimer;
import android.util.Pair;

import com.softwareoverflow.HangTight.helper.WorkoutHelper;

public class WorkoutTimer {

    private CountDownTimer timer;
    private static final int TICK_INTERVAL = 100;
    private long millisecondsRemaining;

    private IWorkoutTimerObserver observer;

    private Workout workout;
    private int prepTimeMillis;

    private int currentRep = 1, currentSet = 1;

    private WorkoutHelper.WorkoutSection currentSection = null;

    private boolean isPaused;

    /**
     * @param workout         The Workout object to construct the timer for.
     * @param prepTimeSeconds The number of seconds to run the prepare countdown for
     */
    public WorkoutTimer(Workout workout, int prepTimeSeconds) {
        // TODO - retrofit preparation time into this
        this.workout = workout;
        this.prepTimeMillis = prepTimeSeconds * 1000;

        createTimer((workout.getDuration() + prepTimeSeconds) * 1000);
    }

    public void start() {
        timer.start();
    }

    public void cancel() {
        timer.cancel();
    }

    /**
     * Skip the current timer section execution
     */
    public void skip() {
        Pair workoutSectionTimePair = WorkoutHelper.getCurrentSection(workout, millisecondsRemaining);
        millisecondsRemaining -= (long) workoutSectionTimePair.second;

        if(isPaused){
            // Allow the UI to be updated if skip is pressed whilst paused
            notifyObservers();
        } else {
            // Cancel and recreate the timer
            timer.cancel();
            createTimer(millisecondsRemaining);
            timer.start();
        }
    }

    /**
     * Pause the current timer execution
     */
    public void pause() {
        isPaused = true;
        timer.cancel();
    }

    /**
     * Resume the current execution
     */
    public void resume() {
        isPaused = false;
        createTimer(millisecondsRemaining);
        timer.start();
    }

    public WorkoutHelper.WorkoutSection getCurrentSection() {
        return currentSection;
    }

    private void createTimer(long durationMillis) {
        timer = new CountDownTimer(durationMillis, TICK_INTERVAL) {
            @Override
            public void onTick(long millisUntilFinished) {
                millisecondsRemaining = millisUntilFinished;
                notifyObservers();
                /*if (observer == null)
                    return;

                millisecondsRemaining = millisUntilFinished;

                Pair workoutSectionTimePair = WorkoutHelper.getCurrentSection(workout, millisUntilFinished);

                if (skipFlag) {
                    millisUntilFinished -= (long) workoutSectionTimePair.second;

                    workoutSectionTimePair = WorkoutHelper.getCurrentSection(workout, millisUntilFinished);
                }

                WorkoutHelper.WorkoutSection workoutSection = (WorkoutHelper.WorkoutSection) workoutSectionTimePair.first;
                if (workoutSection != currentSection) {
                    if (workoutSection == WorkoutHelper.WorkoutSection.PREPARE) {
                        observer.onPrepareStart();
                    } else if (workoutSection == WorkoutHelper.WorkoutSection.HANG) {
                        observer.onHangStart();

                        if (currentSection == WorkoutHelper.WorkoutSection.REST) {
                            observer.onRepComplete(++currentRep);
                        } else if (currentSection == WorkoutHelper.WorkoutSection.RECOVER) {
                            observer.onSetComplete(++currentSet);
                            currentRep = 1;
                        }
                    } else if (workoutSection == WorkoutHelper.WorkoutSection.REST) {
                        observer.onRestStart();
                    } else if (workoutSection == WorkoutHelper.WorkoutSection.RECOVER) {
                        observer.onRecoverStart();
                    }
                }

                currentSection = workoutSection;

                if (skipFlag) {
                    // Cancel and recreate the timer to simulate skipping the section
                    this.cancel();
                    createTimer(millisUntilFinished);
                    timer.start();
                    skipFlag = false;
                } else {
                    observer.onTick(millisUntilFinished, (long) workoutSectionTimePair.second, TICK_INTERVAL);
                }*/
            }

            @Override
            public void onFinish() {
                if (observer != null)
                    observer.onTimerComplete();
            }
        };
    }

    private void notifyObservers(){
        if (observer == null)
            return;

        Pair workoutSectionTimePair = WorkoutHelper.getCurrentSection(workout, millisecondsRemaining);

        WorkoutHelper.WorkoutSection workoutSection = (WorkoutHelper.WorkoutSection) workoutSectionTimePair.first;
        if (workoutSection != currentSection) {
            if (workoutSection == WorkoutHelper.WorkoutSection.PREPARE) {
                observer.onPrepareStart();
            } else if (workoutSection == WorkoutHelper.WorkoutSection.HANG) {
                observer.onHangStart();

                if (currentSection == WorkoutHelper.WorkoutSection.REST) {
                    observer.onRepComplete(++currentRep);
                } else if (currentSection == WorkoutHelper.WorkoutSection.RECOVER) {
                    observer.onSetComplete(++currentSet);
                    currentRep = 1;
                }
            } else if (workoutSection == WorkoutHelper.WorkoutSection.REST) {
                observer.onRestStart();
            } else if (workoutSection == WorkoutHelper.WorkoutSection.RECOVER) {
                observer.onRecoverStart();
            }
        }

        currentSection = workoutSection;
        observer.onTick(millisecondsRemaining, (long) workoutSectionTimePair.second, TICK_INTERVAL);
    }

    public void setObserver(IWorkoutTimerObserver observer) {
        this.observer = observer;
    }
}

File: app/src/main/java/com/softwareoverflow/hangtight/workout/WorkoutTimer.kt
package com.softwareoverflow.hangtight.workout

import com.softwareoverflow.hangtight.helper.WorkoutHelper
import com.softwareoverflow.hangtight.util.ListObjectIterator

class WorkoutTimer(
    workout: Workout,
    private val observer: IWorkoutTimerListener,
    prepTime: Int,
) {
    private var millisecondsRemaining: Long = (workout.duration + prepTime) * 1000L

    private val timerProvider: IWorkoutTimerProvider

    private var isRunning = false
    private var isPaused: Boolean = false

    private val workoutSets = ListObjectIterator(workout.getTimedSections(prepTime).listIterator())
    private var currentSection = workoutSets.next()

    private var millisRemainingInSection = getCurrentSectionTime()

    init {
        timerProvider = WorkoutTimerProvider(
            onTimerFinish = {
                getTimerProvider().cancelTimer()
                observer.onFinish()
            },
            onTimerTick = { tickInterval, millisUntilFinished ->
                observer.onTimeChange(
                    (millisRemainingInSection / 1000).toInt(),
                    (millisecondsRemaining / 1000).toInt()
                )

                if (millisRemainingInSection <= 0 && millisUntilFinished > tickInterval)
                    startNextWorkoutSection()

                millisRemainingInSection -= tickInterval
                millisecondsRemaining -= tickInterval
            }
        )

        getTimerProvider().createTimer(millisecondsRemaining)

        observer.onSectionChange(currentSection)
    }

    fun rewindSection() {
        isRunning = false
        getTimerProvider().cancelTimer()

        val milliSecondsToReset =
            getCurrentSectionTime() - millisRemainingInSection // Amount of time already completed in section being reset
        millisecondsRemaining += milliSecondsToReset
        millisecondsRemaining =
            ((millisecondsRemaining + 999) / 1000) * 1000 // Round up to the nearest second (in millis) to prevent the frequent polling of the timer getting out of sync

        startPreviousWorkoutSection()

        if (!isPaused) {
            getTimerProvider().createTimer(millisecondsRemaining)
            isRunning = true
            getTimerProvider().startTimer()
        }
    }

    /** Skips the current section of the workout **/
    fun skipSection() {
        millisecondsRemaining -= millisRemainingInSection
        millisecondsRemaining =
            ((millisecondsRemaining + 999) / 1000) * 1000 // Round up to the nearest second (in millis) to prevent the frequent polling of the timer getting out of sync

        if (millisecondsRemaining <= 0) {
            getTimerProvider().cancelTimer() // Cancel the timer to prevent onFinish being called multiple times
            observer.onFinish()
            return
        }

        startNextWorkoutSection()

        // Cancel and recreate the timer
        isRunning = false
        getTimerProvider().cancelTimer()

        if (!isPaused) {
            getTimerProvider().createTimer(millisecondsRemaining)
            isRunning = true
            getTimerProvider().startTimer()
        }
    }

    /** Allows the pausing / resuming of the timer**/
    fun togglePause(isPaused: Boolean) {
        this.isPaused = isPaused

        if (isPaused) {
            isRunning = false
            getTimerProvider().cancelTimer()
        } else if (!isRunning) {
            millisecondsRemaining =
                ((millisecondsRemaining + 999) / 1000) * 1000 // Display lags behind by 1s - round up
            millisRemainingInSection =
                ((millisRemainingInSection + 999) / 1000) * 1000// Display lags behind by 1s - round up

            getTimerProvider().createTimer(millisecondsRemaining)

            isRunning = true
            getTimerProvider().startTimer()
        }
    }

    private fun startNextWorkoutSection() {
        var nextSection = workoutSets.tryGetNext()

        if(nextSection?.section == WorkoutHelper.WorkoutSection.PREPARE){
            nextSection = workoutSets.tryGetNext()
        }

        if(nextSection != null) {
            currentSection = nextSection
            millisRemainingInSection = getCurrentSectionTime()

            observer.onSectionChange(currentSection)
            observer.onTimeChange(
                millisRemainingInSection.toInt() / 1000,
                millisecondsRemaining.toInt() / 1000
            )
        }
    }

    private fun startPreviousWorkoutSection() {
        val previousSection = workoutSets.tryGetPrevious()
        if(previousSection != null) {
            currentSection = previousSection
            millisecondsRemaining += getCurrentSectionTime()
            observer.onSectionChange(currentSection)
        }

        millisRemainingInSection = getCurrentSectionTime()
        observer.onTimeChange(
            millisRemainingInSection.toInt() / 1000,
            millisecondsRemaining.toInt() / 1000
        )
    }

    fun start() {
        isRunning = true
        getTimerProvider().startTimer()
    }

    fun cancel() {
        getTimerProvider().cancelTimer()
    }

    private fun getCurrentSectionTime() = currentSection.durationSeconds * 1000L

    private fun getTimerProvider() = timerProvider
}
